Polio 619 refactor budget front end

quang-le opened the pull request:
Refactor Front-end of the budget tool to accomodate the use of the workflow engine

## Self proof reading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented
- [ ] Are my typescript files well typed - half-half
- [X] Did I add translations
- [X] My migrations file are included
- [ ] Are there enough tests - Not in the front
## Changes

###  Features
- Add step button + modal has been replaced by buttons generated dynamically according to the workflow state
- The Map in Budget details has been removed to give more space to the Table, which allowed to remove the `BudgetFiles`modal in the desktop version: the files and links are shown directly in the table
- Budget Steps are not editable anymore
- The notion of deletion/restoration of events has been replaced by show/hide, which don't open a modal anymore.
- Filters by team and recipient have been removed as they don't make sense anymore in the context of the steps being constrained by the workflow.
- When adding links, users can now choose an alias, that will be displayed in the table
- Step creation  buttons can have 3 colors: "primary", "red" or "green". These are determined in the workflow

### In the code

- Data for the Budget Table is received from a single endpoint from the backend. No values are computed in the front-end (as opposed to previously where e.g. status was computed from all events for the campaign)
- All calls to `/api/polio/campaign` have been removed
- `CreateEditBudgetEvent`has been replaced with `CreateBudgetStep` and `CreateOverrideStep`
- Table and Card layouts for BudgetDetails have been split and moved to their own component
- Budget status and step names come directly from the backend and are not translated in the front-end
- Added `AddMultipleLinks` and `NamedLink`components for managing links' aliases
- Most calls to `api/profiles` have been removed since the backend handles  the necessary checks
- links are now treated as a list of objects with an `alias` and a `url`
- likewise have a `filename` that we can display in the table. 
- Step creation button colors are handled with classNames, which explains why the are constrained to predetermined values


## How to test

- Get the workflow config file from [Google Drive](https://drive.google.com/drive/u/0/folders/1ua4yzyFOGayweQVNKP0hyBNDKhQrPlx1)
- Go to Django admin -> workflow models and paste the json from the file there
- To test, you will need need the following teams with the corresponding team ids: 
    - 1: RRT
    - 2: ORPG Ops
    - 3: ORPG wider group
    - 4: UNICEF
    - 5: WHO
    - 6: MoH (stands for Ministry of Health)
    - 7: GPEI coordinator
 - If the team ids are not free, you will need to change the workflow config with the ids you will use. If you're using a polio DB it should work since no teams were created there yet
 - You will then need a user for each team. Ideally you need to make separate users, otherwise you will not be able to test that the actions are restricted based on the team
 - Once all this is set up, you can test the workflow by selecting a campaign and starting to create steps. The "scenario" of the workflow is in [Whimsical](https://whimsical.com/latest-version-budget-workflows-RkEXevU68FbkjJScFzJU2w)

## Print screen / video

https://user-images.githubusercontent.com/38907762/194546197-392f87e7-2423-41b3-987d-374ca4a7fc6f.mov


## Notes

⚠️ Includes [POLIO-617](https://github.com/BLSQ/iaso/tree/POLIO-617_BWE_backend) changes (backend)

It's a big PR since it incorporates current work of @olethanh on the backend, so:
@niconoe Can you please have a look at the backend part
@beygorghor Can you check the front

Both are quite big in their own right, though @beygorghor  you'll see that a lot of the front-end changes are renaming stuff or replacing types

Known issues:

- Modal fields for each step are not yet dynamic ("displayed fields")
- Backend for override feature is not ready
- Display bug with next steps when there are several possible options

madewulf commented:
Sorry, I fixed a bug with projects api in the middle of everything. That's a bit messy.

madewulf commented:
Thanks @pjaspers and @mestachs for the feedbacks.

olethanh committed:
adapt model

olethanh committed:
Read API

olethanh committed:
Start api to transition

quang-le committed:
POLIO-619: fix typo

quang-le committed:
POLIO-619: update routes

quang-le committed:
POLIO-619: add mockAPI calls

quang-le committed:
POLIO-619: add new types

quang-le committed:
POLIO-619: update table configs

quang-le committed:
POLIO-619: update index.tsx

quang-le committed:
POLIO-619: create new component to create step

quang-le committed:
POLIO-619: update BudgetDetails

quang-le committed:
POLIO-619: mobile layout for budgetDetails

olethanh committed:
Improve DeletionFilterBackend

olethanh committed:
POLIO-621 Fix audit log for multi scope and vacine tab

olethanh committed:
Add /budgetsteps/ api

olethanh committed:
Add budget_last_updated_at on campaigns

olethanh committed:
rename model budgetstepfile add filename field

to keep the original file name

olethanh committed:
save Files uploaded, return them nicely

olethanh committed:
commend on budget is not required

quang-le committed:
POLIO-619: adapt Budget mobile layout to API

quang-le committed:
POLIO-619: cleanup code
- remove unused type
- add key to component array
- remove imports

olethanh committed:
Improve doc

quang-le committed:
POLIO-619: add campaignId

quang-le committed:
POLIO-619: update validation

olethanh committed:
filter for user

olethanh committed:
add tests

olethanh committed:
Remove budgetStep.links temp

olethanh committed:
Make links works in api

quang-le committed:
POLIO-619: add params to budet details route

quang-le committed:
POLIO-619: replace opo ups with CreateBudgetStep

quang-le committed:
POLIO-619: add override step

quang-le committed:
POLIO-619: update mock data to fix crash

quang-le committed:
POLIO-619: update form validation (WIP)

- remove obsolete fields
- add validation for override

quang-le committed:
POLIO-619: update BudgetDetails

- update CreateBudgetStep component
- use CreateOverride
- split component for readability

quang-le committed:
POLIO-619: remove unused method

quang-le committed:
POLIO-619: add translations

olethanh committed:
Transition add color field

olethanh committed:
Add more test and fix bugs

olethanh committed:
make BudgetStep soft deletable

olethanh committed:
clean

olethanh committed:
Add color in test workflow

quang-le committed:
POLIO-619: add translations

quang-le committed:
POLIO-619: add links with alias

olethanh committed:
Add MailTemplate

quang-le committed:
POLIO-619: add Multiple links component

quang-le committed:
Merge remote-tracking branch 'origin/POLIO-617_BWE_backend' into POLIO-619_refactor_budget-front-end

olethanh committed:
MailTemplate add field for subject

olethanh committed:
Fix url in browsable api

olethanh committed:
Correctly deleted_at on BudgetStep

quang-le committed:
Merge branch 'POLIO-617_BWE_backend' of https://github.com/BLSQ/iaso into POLIO-619_refactor_budget-front-end

olethanh committed:
MailTemplate missing migration

quang-le committed:
Merge branch 'POLIO-617_BWE_backend' of https://github.com/BLSQ/iaso into POLIO-619_refactor_budget-front-end

olethanh committed:
Remove duplicate code; fix swagger generation

olethanh committed:
Add budget.possible_state as per new api

olethanh committed:
Add user and team info

quang-le committed:
Merge branch 'POLIO-617_BWE_backend' of https://github.com/BLSQ/iaso into POLIO-619_refactor_budget-front-end

olethanh committed:
Add admin for BudgetStep (and file and link)

quang-le committed:
Merge branch 'POLIO-617_BWE_backend' of https://github.com/BLSQ/iaso into POLIO-619_refactor_budget-front-end

quang-le committed:
POLIO-619: update blsq-comp

quang-le committed:
POLIO-617: replace mock calls with API calls

quang-le committed:
POLIO-619: adapt mobile layout

quang-le committed:
POLIO-619: fix typing

quang-le committed:
POLIO-619: change color for feedback step

quang-le committed:
POLIO-619: comment out validation code

quang-le committed:
POLIO-619: remove obsolete components

quang-le committed:
POLIO-619: move and rename files

quang-le committed:
POLIO-619: fix webpack build

quang-le committed:
POLIo-619: fiix button color

olethanh committed:
start email workflow

quang-le committed:
POLIO-619:filter budget steps by campaign

quang-le committed:
POLIO-619: cleanup Budget folder

quang-le committed:
POLIO-619: add types, cleanup comments

olethanh committed:
Fix uploading of file in Budget

quang-le committed:
POLIO-619 add missing import

quang-le committed:
Merge branch 'POLIO-619_refactor_budget-front-end' of https://github.com/BLSQ/iaso into POLIO-619_refactor_budget-front-end

olethanh committed:
fix default order

quang-le committed:
POLIO-619: fix refresh on step save

quang-le committed:
Merge branch 'POLIO-619_refactor_budget-front-end' of https://github.com/BLSQ/iaso into POLIO-619_refactor_budget-front-end

quang-le committed:
POLIO-619: fix table pagination bug on mobile

quang-le committed:
POLIO-619: fix links input flicker

quang-le committed:
POLIo-619: resize link alias field

quang-le committed:
POLIO-619: remove obsolote fiels, improve typing

quang-le committed:
POLIO-619: add APi call for show/hide step

olethanh committed:
fix sending links

olethanh committed:
Add plain text template field

olethanh committed:
Create an Email with both the plain text and html

olethanh committed:
Add proper txt template, imporove html template

olethanh committed:
fix link generation

olethanh committed:
send the e-mails

olethanh committed:
Merge branch 'POLIO-619_refactor_budget-front-end' into POLIO-617_BWE_backend

olethanh committed:
Merge remote-tracking branch 'origin/main' into POLIO-619_refactor_budget-front-end

olethanh committed:
Fix MyPy

olethanh committed:
Correct requirements.txt

olethanh committed:
Merge branch 'POLIO-619_refactor_budget-front-end' into POLIO-617_BWE_backend

olethanh committed:
Fix default order of BudgetStep

quang-le committed:
POLIO-619: add virus notification column

- add field to serializer
- add column to config
- pass sorting param to API

quang-le committed:
POLIO-619 resize files modal

olethanh committed:
Move Workflow to the database

olethanh committed:
Add cache for get_workflow

olethanh committed:
Add a proper label for budget with no campaign

olethanh committed:
Add possible_transition on Budget

olethanh committed:
Fix swagger definition for buget api

olethanh committed:
implement checking the attachment file

olethanh committed:
/budget/api allow dynamically requesting the fields

olethanh committed:
Make softDelete works

olethanh committed:
Allow filtering BudgetStep by transition

quang-le committed:
POLIO-619: get required fields from workflow

quang-le committed:
Merge branch 'POLIO-617_BWE_backend' of https://github.com/BLSQ/iaso into POLIO-619_refactor_budget-front-end

olethanh committed:
missing migration files

quang-le committed:
Merge branch 'POLIO-617_BWE_backend' of https://github.com/BLSQ/iaso into POLIO-619_refactor_budget-front-end

olethanh committed:
Proper use of permission

olethanh committed:
allow ordering and filtering by state

olethanh committed:
Ordering by status in frontend

quang-le committed:
Merge branch 'POLIO-617_BWE_backend' of https://github.com/BLSQ/iaso into POLIO-619_refactor_budget-front-end

quang-le committed:
POLIO-619: restore filters

quang-le committed:
POLIO-619: add validation for "attachment"

quang-le committed:
POLIO-619: update message

quang-le committed:
POLIO-619: add required and error to files, links

quang-le committed:
POLIO-619: improve error mgt for links

olethanh committed:
Fix email sending

quang-le committed:
POLIO-619: increase length of transition key

quang-le committed:
POLIO-619: small bug fixes in serializer

- add comment field
- fix condition

quang-le committed:
POLIO-619: fix url when coming from email

quang-le committed:
POLIO-619: update bluesquare-components

quang-le committed:
POLIO-619: change translation

olethanh committed:
fix color in button email

olethanh committed:
fix typo in comment

olethanh committed:
Fix filtering Step and Transition with the same label

olethanh committed:
In mail group files and links together

quang-le committed:
POLIO-619: add last update column

quang-le committed:
POLIO-619: fix links for quicktransition

- files from previous steps are converted to links in the form
- So the user can add other files without losibg files from prev step
- form is directly sendable when coming from quickTransition

niconoe commented on hat/api/token_authentication.py:
```diff
@@ -27,9 +28,12 @@ def generate_auto_authentication_link(link, user):
+    protocol = "https" if settings.SSL_ON else "http"
+
+    final_link = "%s://%s/api/token_auth/?token=%s&next=%s" % (protocol, domain, access_token, encoded_link)
+    print(final_link)
```
Forgotten from debugging?

niconoe commented on plugins/polio/admin.py:
```diff
@@ -75,6 +78,42 @@ class CampaignGroupAdmin(admin.ModelAdmin):
+            d = json.dumps(obj.definition, indent=2)
+            html = f"<pre>{d}</pre>"
+            return mark_safe(html)
+        except Exception as e:
+            print(e)
```
Maybe we can catch a less broad exception, and use the logging module rather than print().

quang-le committed:
POLIO-619: improve labels and translations

quang-le committed:
POLIO-619: stop displaying inactive buttons

quang-le committed:
POLIO-619: add translations

quang-le committed:
POLIO-619 display next budget steps

quang-le committed:
POLIO-619: fix table sort bug

olethanh committed:
Rename attached files in attachment in front

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetailsFilters.tsx:
```diff
@@ -0,0 +1,79 @@
+            params,
+            saveSearchInHistory: false,
+        });
+    // const { data: teams, isFetching } = useGetTeamsDropDown();
```
this can be removed

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetailsTableLayout.tsx:
```diff
@@ -0,0 +1,74 @@
+    params,
+    columns,
+    isFetching,
+    resetPageToOne,
```
this is not required anymore since the last changes i did on the table (normally)

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/mobile/BudgetDetailsFiltersMobile.tsx:
```diff
@@ -0,0 +1,35 @@
+
+export const BudgetDetailsFiltersMobile: FunctionComponent<Props> = ({
+    params,
+    // expand,
```
to remove

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetFilters.tsx:
```diff
@@ -25,28 +21,21 @@ type Props = {
     const theme = useTheme();
     const isXSLayout = useMediaQuery(theme.breakpoints.down('xs'));
-    const isSmLayout = useMediaQuery(theme.breakpoints.down('sm'));
+    // const isSmLayout = useMediaQuery(theme.breakpoints.down('sm'));
```
to remove

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetFilters.tsx:
```diff
@@ -64,14 +53,14 @@ export const BudgetFilters: FunctionComponent<Props> = ({
                     />
                 </Grid>
-                <Grid item xs={12} sm={12} md={6}>
+                {/* <Grid item xs={12} sm={12} md={6}>
```
to remove

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetStatus.tsx:
```diff
@@ -57,13 +38,15 @@ export const BudgetStatus: FunctionComponent<Props> = ({ budgetStatus }) => {
+            {budgetStatus && (
+                <Typography
+                    variant="h5"
+                    className={`${classes.title}`}
```
`className={classes.title}`

beygorghor commented on plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateBudgetStep.tsx:
```diff
@@ -0,0 +1,297 @@
+    const { mutateAsync: saveBudgetStep } = useSaveBudgetStep();
+    const classes = useStyles();
+    const dispatch = useDispatch();
+    //@ts-ignore
```
this is not necessary no ?

quang-le replied:
it seems my ts-linter is on drugs or something,

beygorghor commented on plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateBudgetStep.tsx:
```diff
@@ -0,0 +1,297 @@
+    //@ts-ignore
+    const quickTransition = params?.quickTransition
+    const onSuccess = useCallback(() => {
+        const trimmedParams = {...params}
```
you have some eslint issues here

beygorghor commented on plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateBudgetStep.tsx:
```diff
@@ -0,0 +1,297 @@
+                                required={requiredFields.includes('links')}
+                            />
+                        </Box>
+                        {/* @ts-ignore */}
```
weird

quang-le replied:
Object is possibly undefined, says tslint. I don't see how though

beygorghor commented on plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateBudgetStep.tsx:
```diff
@@ -0,0 +1,297 @@
+            >
+                {userHasTeam && (
+                    <>
+                        <>
```
you can remove this fragment

beygorghor commented on plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateBudgetStep.tsx:
```diff
@@ -0,0 +1,297 @@
+                                ))}
+                            </>
+                        )}
+                        {/* @ts-ignore */}
```
useless

beygorghor commented on plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx:
```diff
@@ -0,0 +1,234 @@
+                                }}
+                                value={values.type}
+                                errors={getErrors('type')}
+                                label={MESSAGES.eventType}
```
this translation is missing

beygorghor commented on plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx:
```diff
@@ -0,0 +1,234 @@
+    });
+
+    const titleMessage = transitionLabel;
+    // const titleMessage = MESSAGES.newBudgetStep;
```
to remove

beygorghor commented on plugins/polio/js/src/pages/Budget/CreateBudgetStep/CreateOverrideStep.tsx:
```diff
@@ -0,0 +1,234 @@
+            >
+                {userHasTeam && (
+                    <>
+                        <>
```
useless fragment

beygorghor commented on plugins/polio/js/src/pages/Budget/MultipleLinks/AddMultipleLinks.tsx:
```diff
@@ -0,0 +1,101 @@
+    const { formatMessage } = useSafeIntl();
+    const {
+        setFieldValue,
+        // @ts-ignore
```
useless

beygorghor commented on plugins/polio/js/src/pages/Budget/MultipleLinks/AddMultipleLinks.tsx:
```diff
@@ -0,0 +1,101 @@
+    } = useFormikContext();
+
+    // @ts-ignore
+    const links = useMemo(() => values?.links ?? [], [values?.links]);
```
if you type values you can remove the ignore

quang-le replied:
Yep, I was lazy...

beygorghor commented on plugins/polio/js/src/pages/Budget/NextBudgetStep.tsx:
```diff
@@ -0,0 +1,49 @@
+});
+
+// @ts-ignore
+const useStyles = makeStyles(styles);
```
`
const useStyles = makeStyles(theme => ({
    title: {
        fontWeight: 'bold',
        [theme.breakpoints.down('md')]: {
            fontSize: 18,
        },
    },
    spacer: { marginRight: '5px' },
}));
`
is fixing the error, don't ask me why

beygorghor commented on plugins/polio/js/src/pages/Budget/hooks/api/useGetBudgetDetails.ts:
```diff
@@ -0,0 +1,54 @@
+    return getRequest(`/api/polio/budgetsteps/?${queryString}`);
+};
+
+export const useGetBudgetDetails = params => {
```
no typing here ?

beygorghor commented on plugins/polio/js/src/pages/Budget/hooks/api/useGetStatusDropDown.ts:
```diff
@@ -0,0 +1,38 @@
+import { waitFor } from '../../../../../../../../hat/assets/js/apps/Iaso/utils';
+
+const getSteps = async () => {
+    await waitFor(1000);
```
Why ?

quang-le replied:
Actually, it's not used anymore, I'll delete it. Leftover from when the api wasn't ready

beygorghor commented on plugins/polio/js/src/pages/Budget/utils.tsx:
```diff
@@ -52,32 +50,41 @@ export const extractFileName = (fileUrl: string): string => {
-    const linksArray = links.split(',');
-    return linksArray.map((link, index) => {
-        const trimmedLink = link.trim();
+    // const linksArray = links.split(',');
```
to remove

beygorghor commented on plugins/polio/js/src/pages/Budget/utils.tsx:
```diff
@@ -52,32 +50,41 @@ export const extractFileName = (fileUrl: string): string => {
+export const makeFileLinks = (files: FileWithName[]): React.ReactNode => {
     return files.map((file, index) => {
-        const fileName = extractFileName(file.file) || file.file;
+        // const fileName = extractFileName(file) || file;
```
to remove

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetails.tsx:
```diff
@@ -0,0 +1,333 @@
+    }, [campaignId, rest, showHidden]);
+
+    const checkBoxLabel = formatMessage(MESSAGES.showHidden);
+    // @ts-ignore
```
you can type a state

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetails.tsx:
```diff
@@ -0,0 +1,333 @@
+            campaign_id: campaignId,
+            transition_key__in: transition_key,
+        };
+    }, [campaignId, rest, showHidden]);
```
transition_key is not in the deps

quang-le replied:
well spotted. My linter was on strike yesterday apparently

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetails.tsx:
```diff
@@ -0,0 +1,333 @@
+        const override = budgetInfos?.next_transitions?.find(
+            transition => transition.key === 'override',
+        );
+        const toDisplay = new Set(regular?.filter(transition=>transition.key!==budgetInfos?.current_state?.key).map(transition=>transition.label));
```
eslint where are you ?

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetails.tsx:
```diff
@@ -0,0 +1,333 @@
+        );
+        const toDisplay = new Set(regular?.filter(transition=>transition.key!==budgetInfos?.current_state?.key).map(transition=>transition.label));
+        return { regular, override, toDisplay };
+    }, [budgetInfos?.next_transitions]);
```
budgetInfos?.current_state?.key ?

quang-le replied:
my linter is letting me down these days

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetails.tsx:
```diff
@@ -0,0 +1,333 @@
+                    }
+                }}
+            />
+            {/* @ts-ignore */}
```
useless

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetails.tsx:
```diff
@@ -0,0 +1,333 @@
+            {/* @ts-ignore */}
+            <Box
+                // @ts-ignore
+                className={`${classes.containerFullHeightNoTabPadded}`}
```
`className={classes.containerFullHeightNoTabPadded}`

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetails.tsx:
```diff
@@ -0,0 +1,333 @@
+                                                                isMobileLayout={
+                                                                    isMobileLayout
+                                                                }
+                                                                // displayedFields={step.displayed_fields}
```
to remove

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetails.tsx:
```diff
@@ -0,0 +1,333 @@
+                                                                    label: step.label,
+                                                                    color: step.color,
+                                                                    disabled:
+                                                                        !step.allowed,
```
to remove

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetails.tsx:
```diff
@@ -0,0 +1,333 @@
+                                                    isMobileLayout={
+                                                        isMobileLayout
+                                                    }
+                                                    // displayedFields={step.displayed_fields}
```
to remove

beygorghor commented on plugins/polio/js/src/pages/Budget/BudgetDetails/BudgetDetails.tsx:
```diff
@@ -0,0 +1,333 @@
+                                                        color: nextSteps
+                                                            .override.color,
+
+                                                        // tooltipText={step.reason_not_allowed}
```
to remove

olethanh committed:
missing migration

beygorghor reviewed (changes requested):
Huuuuge job  !! 👏 
- I have an issue while requesting a budget and leaving a second link empty 
![Screenshot 2022-10-18 at 11 22 01](https://user-images.githubusercontent.com/12494624/196391301-0410c94a-6d4f-481e-a070-58b557841c36.png)
- i'm not a huge fan of the error set by default if you touch another field, also files should be required here:

![Screenshot 2022-10-18 at 11 21 53](https://user-images.githubusercontent.com/12494624/196391305-6afe
![Screenshot 2022-10-18 at 11 23 26](https://user-images.githubusercontent.com/12494624/196391680-1ffcebca-6d4e-490a-984d-e1beb34b9383.png)
90c8-87f5-471f-85ba-19f3c04021cc.png)

quang-le committed:
POLIO-619: remove bad migration

quang-le committed:
POLIO-619: code review

quang-le commented:
> Huuuuge job !! 👏
> 
>     * I have an issue while requesting a budget and leaving a second link empty
>       ![Screenshot 2022-10-18 at 11 22 01](https://user-images.githubusercontent.com/12494624/196391301-0410c94a-6d4f-481e-a070-58b557841c36.png)
> 
>     * i'm not a huge fan of the error set by default if you touch another field, also files should be required here:
> 
> 
> ![Screenshot 2022-10-18 at 11 21 53](https://user-images.githubusercontent.com/12494624/196391305-6afe ![Screenshot 2022-10-18 at 11 23 26](https://user-images.githubusercontent.com/12494624/196391680-1ffcebca-6d4e-490a-984d-e1beb34b9383.png) 90c8-87f5-471f-85ba-19f3c04021cc.png)

The error I spun in another ticket, since we're deploying for demo, and not for client use. The thing with the error and required fields we can discuss, as it's one of links or files that is required, not both, so it's a bit trickier to signal to the user with our usual conventions

quang-le committed:
update package lock

quang-le commented:
The package-lock conflict seems to be with bluesquare-components. The safest bet is probaby to update bluesquare-components on main, since resolving package-lock conflicts by hand is both tedious and error prone

quang-le committed:
update dep to blsq-comp

quang-le committed:
Merge branch 'update_package_lok' into POLIO-619_refactor_budget-front-end

olethanh committed:
Fix test broken because of missing permission

olethanh committed:
fix typing
